### DISCUSSION TITLE
PicoGK Viewer

### DISCUSSION THREAD

**Discussion by AasimRashid** [leap71/PicoGK#27] [Q&A] [answered]
Hello,

I plan to develop a UI for the PicoGK. For that purpose I am working with Blazor in Visual Studio. The first challenge is to embed PicoGK Viewer window inside webview. In my current implementation, PicoGK Viewer opens in a separate window. I was wondering if it is possible that PicoGK Viewer opens inside a Blazor component (i.e. viewer is embedded in the web browser )? Any guide or hint is welcome. Thanks!

**Comment by LinKayser** [comment, marked as answer]
This is not going to be trivial. The PicoGK viewer uses GLFW to provide access to a generic operating system window. You could see, if you could somehow gain access to a Window handle of your Blazor compoinent, provided, it actually uses an OS window (and not its own windowing system inside the browser window, which I suspect). If you can get access to the OS window handle, you could attempt to child the GLFW window to this handle. Again, not a trivial undertaking.

The other thing you could do is get rid of the PicoGK viewer and re-implement it using WebGL, rendering the triangles. There's a lot of functionality in the viewer, but most of it is implemented outside the C++ part. To replace the PicoGK viewer you would need to pass keystrokes, mouse movements, render triangle meshes, render polylines, set the global view matrix, and be able to set the individual group matrices of each viewer group (which is used for animated objects).

Have a look at the C++ code for the viewer here: https://github.com/leap71/PicoGKRuntime/blob/main/Source/PicoGKGLViewer.h

It is not super complex, because I tried to keep it quite generic, but it's also not trivial.

What I am currently looking at is using Avalonia for a viewer framework.

Best,
Lin

**Reply by AasimRashid** [reply]
Thank you very much for the quick and detailed reply.

I have already tried the first method (albeit half heartedly) that you mentioned but could not figure it out. I was trying to follow the procedure which i learned in a tutorial for Windows Form application. 

I have a gut feeling that a new viewer needs to be implemented using WebGL. However, I do not have the sufficient knowledge and skills to do such implementation at the moment. My plan is to learn more in this domain and do the implementation after a couple of months.

Thanks again for your elaboration. 

Kind regards
Asim

**Comment by AasimRashid** [comment]
I had not opportunity to work with Viewer, however, I have developed a few dialog boxes. Here is how Cylinder dialog box looks. 


https://github.com/user-attachments/assets/97df7226-f9f3-4459-ab5c-83fa78be3df1



Cheers
Asim

**Reply by Mech-eng-au** [reply]
Wow, that looks great! And such an easy way to iterate with different parameters without scrolling through code. Great job!

**Reply by LinKayser** [reply]
> I had not opportunity to work with Viewer, however, I have developed a few dialog boxes. Here is how Cylinder dialog box looks.
> 
>  blazor.mp4 
> Cheers Asim

Could you explain a bit what you did? Are you running PicoGK as a headless process in the background, exporting an STL mesh, which you load and show in the viewer? Or is it more complex than that?

Best,
Lin

**Reply by AasimRashid** [reply]
You are right! PicoGK is running in headless mode and after export, it is being displayed in a WebGL viewer. This is the simplest that i could figure out to begin with.

All the best
Asim

**Reply by Mech-eng-au** [reply]
@AasimRashid, is it possible for you to share the code? :)

**Reply by rawwerks** [reply]
@AasimRashid - this is very cool! i would also be excited to learn from your example if you'd be willing to share any more tips for how to get the output of picogk into the browser.

**Reply by AasimRashid** [reply]
I am using Microsoft Blazor to build this GUI. I deployed Blazor3d (https://github.com/HomagGroup/Blazor3D) plugin to view the models inside the browser.
